### PR TITLE
*: write binlog to unix socket.

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @disksing @iamxy @qiuyesuifeng PTAL

**Comment by disksing** [comment]
Do we need a switch to turn on/off this feature? Not all user need it.

File: infoschema/infoschema.go [hunk 1]
```diff
@@ -22,10 +22,11 @@ import (
 	"github.com/pingcap/tidb/meta/autoid"
 	"github.com/pingcap/tidb/model"
 	"github.com/pingcap/tidb/mysql"
-	"github.com/pingcap/tidb/perfschema"
+	//"github.com/pingcap/tidb/perfschema"
 	"github.com/pingcap/tidb/table"
 	"github.com/pingcap/tidb/terror"
 	// import table implementation to init table.TableFromMeta
+	"github.com/pingcap/tidb/perfschema"
```
Comment (disksing) [inline infoschema/infoschema.go]: Any special reason to move the import?

File: session_test.go [hunk 2]
```diff
@@ -2554,3 +2556,68 @@ func (s *testSessionSuite) TestSelectHaving(c *C) {
...
+	c.Assert(deletedRow1, DeepEquals, types.MakeDatums(1, 3))
+	deletedRow2, _ := codec.Decode(bin.Mutations[0].DeletedRows[1], 2)
+	c.Assert(deletedRow2, DeepEquals, types.MakeDatums(2, 3))
+}
+
+func getLatestBinlog(c *C, store kv.Storage) *binlog.Binlog {
+	txn, err := store.Begin()
+	c.Assert(err, IsNil)
+	it, err := txn.Seek([]byte("b"))
+	c.Assert(errors.ErrorStack(err), Equals, "")
```
Comment (disksing) [inline session_test.go]: How about `c.Assert(err, IsNil)`?

Comment (coocood) [inline session_test.go]: When error did occur, ErrorStack will print the stack trace.

File: executor/executor_test.go [hunk 3]
```diff
@@ -31,6 +31,7 @@ import (
 	"github.com/pingcap/tidb/kv"
 	"github.com/pingcap/tidb/model"
 	"github.com/pingcap/tidb/parser"
+	"github.com/pingcap/tidb/sessionctx/schemaverion"
```
Comment (hanfei1991) [inline executor/executor_test.go]: verion->version

**Comment by coocood** [comment]
@disksing @hanfei1991 PTAL

File: table/tables/tables.go [hunk 4]
```diff
@@ -515,7 +525,38 @@ func (t *Table) RemoveRecord(ctx context.Context, h int64, r []types.Datum) erro
 	if err != nil {
 		return errors.Trace(err)
 	}
-
+	if binloginfo.Enable {
```
Comment (zimulala) [inline table/tables/tables.go]: Here can use a function.

File: sessionctx/binloginfo/binloginfo.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+)
+
+var Enable = true
+
+// keyType is a dummy type to avoid naming collision in context.
+type keyType int
+
+// String defines a Stringer function for debugging and pretty printing.
+func (k keyType) String() string {
+	return "schema_version"
```
Comment (zimulala) [inline sessionctx/binloginfo/binloginfo.go]: Does it need to judge keyType value?

Comment (disksing) [inline sessionctx/binloginfo/binloginfo.go]: Maybe better to define them as different types.
```
type schemaVersionKey int
type binlogKey int
```

File: sessionctx/binloginfo/binloginfo.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package binloginfo
```
Comment (shenli) [inline sessionctx/binloginfo/binloginfo.go]: How about rename it to binlog?

File: store/localstore/kv.go [hunk 7]
```diff
@@ -20,6 +20,7 @@ import (
 	"sync"
 	"time"
 
+	"encoding/binary"
```
Comment (shenli) [inline store/localstore/kv.go]: This should be with standard libs.

File: table/tables/tables.go [hunk 8]
```diff
@@ -515,7 +525,44 @@ func (t *Table) RemoveRecord(ctx context.Context, h int64, r []types.Datum) erro
...
+		err = t.addDeleteBinlog(ctx, h, r)
+	}
+	return errors.Trace(err)
+}
 
+func (t *Table) addDeleteBinlog(ctx context.Context, h int64, r []types.Datum) error {
+	mutation := t.getMutation(ctx)
+	if t.meta.PKIsHandle {
+		mutation.DeletedIds = append(mutation.DeletedIds, h)
```
Comment (shenli) [inline table/tables/tables.go]: Return nil here, so we could save the else.

**Comment by shenli** [comment]
For same statements, we  use another sql to do some sub jobs. For example, set global variable, it will run an update statement. Do we need to avoid add an update binlog when set global variable?

**Comment by coocood** [comment]
@iamxy @shenli @zimulala PTAL

File: sessionctx/binloginfo/binloginfo.go (hunk 6 above)
Comment (coocood) [inline sessionctx/binloginfo/binloginfo.go]: Then the package name will conflict with `tipb/go-binlog`.

**Comment by coocood** [comment]
@iamxy @shenli @zimulala PTAL

File: session.go [hunk 9]
```diff
@@ -184,18 +185,29 @@ func (s *session) finishTxn(rollback bool) error {
 	if s.txn == nil {
 		return nil
 	}
+	sessVar := variable.GetSessionVars(s)
```
Comment (zimulala) [inline session.go]: `sessVar` seems useless.

File: store/tikv/txn_committer.go [hunk 10]
```diff
@@ -338,35 +341,122 @@ func (c *txnCommitter) Commit() error {
...
+		binRollback := &binlog.Binlog{
+			Tp:          binlog.BinlogType_Rollback,
+			StartTs:     int64(c.startTS),
+			PrewriteKey: c.keys[0],
+		}
+		rollbackData, _ := binRollback.Marshal()
+		req := &binlog.WriteBinlogReq{ClusterID: binloginfo.ClusterID, Payload: rollbackData}
+		resp, err := binloginfo.PumpClient.WriteBinlog(context.Background(), req)
+		if err != nil {
+			log.Error(err)
```
Comment (zimulala) [inline store/tikv/txn_committer.go]: Add more log information. The same as the line 431.

File: store/tikv/txn_committer.go [hunk 11]
```diff
@@ -338,35 +341,122 @@ func (c *txnCommitter) Commit() error {
...
+			}
+			prewriteData, _ := binPrewrite.Marshal()
+			req := &binlog.WriteBinlogReq{ClusterID: binloginfo.ClusterID, Payload: prewriteData}
+			var resp *binlog.WriteBinlogResp
+			resp, err = binloginfo.PumpClient.WriteBinlog(context.Background(), req)
+			if err == nil && resp.Errmsg != "" {
+				err = errors.New(resp.Errmsg)
+			}
+		}
+		ch <- err
```
Comment (zimulala) [inline store/tikv/txn_committer.go]: err -> errors.Trace(err) ?

**Comment by coocood** [comment]
@shenli @zimulala PTAL

File: store/tikv/txn_committer.go [hunk 12]
```diff
@@ -338,35 +341,129 @@ func (c *txnCommitter) Commit() error {
...
+			log.Error(err)
+		}
+		if resp != nil && resp.Errmsg != "" {
+			log.Error(resp.Errmsg)
+		}
+	}()
+}
+
+func (c *txnCommitter) writeCommitBinlog() {
+	if binloginfo.PumpClient == nil {
```
Comment (shenli) [inline store/tikv/txn_committer.go]: Use shouldWriteBinlog?

File: store/tikv/txn_committer.go [hunk 13]
```diff
@@ -338,35 +341,129 @@ func (c *txnCommitter) Commit() error {
...
+				err = errors.New(resp.Errmsg)
+			}
+		}
+		ch <- errors.Trace(err)
+	}()
+	return ch
+}
+
+func (c *txnCommitter) rollbackBinlog() {
+	if binloginfo.PumpClient == nil {
```
Comment (shenli) [inline store/tikv/txn_committer.go]: shouldWriteBinlog?

File: store/tikv/txn_committer.go [hunk 14]
```diff
@@ -338,35 +341,129 @@ func (c *txnCommitter) Commit() error {
...
 			return errors.Trace(err)
 		}
 		log.Warnf("txn commit succeed with error: %v, tid: %d", err, c.startTS)
 	}
+	c.writeCommitBinlog()
 	return nil
 }
 
+func (c *txnCommitter) prewriteBinlog() chan error {
+	if binloginfo.PumpClient == nil {
```
Comment (shenli) [inline store/tikv/txn_committer.go]: shouldWriteBinlog?

File: tidb-server/main.go [hunk 15]
```diff
@@ -91,14 +98,36 @@ func main() {
...
+	if err != nil {
+		log.Fatal(errors.ErrorStack(err))
+	}
+	if cluster := u.Query().Get("cluster"); cluster != "" {
+		id, err1 := strconv.ParseUint(cluster, 10, 64)
+		if err1 != nil {
+			log.Fatal(errors.ErrorStack(err1))
+		}
+		binloginfo.ClusterID = id
+	}
```
Comment (zimulala) [inline tidb-server/main.go]: Can line 101-112 be extract into a function？I think  simpler and shorter code in the function of main is better. The same as line 122-130.

File: tidb-server/main.go [hunk 16]
```diff
@@ -91,14 +98,36 @@ func main() {
 	printer.PrintTiDBInfo()
 	log.SetLevelByString(cfg.LogLevel)
 
-	store, err := tidb.NewStore(fmt.Sprintf("%s://%s", *store, *storePath))
+	fullPath := fmt.Sprintf("%s://%s", *store, *storePath)
+	u, err := url.Parse(fullPath)
+	if err != nil {
+		log.Fatal(errors.ErrorStack(err))
+	}
+	if cluster := u.Query().Get("cluster"); cluster != "" {
```
Comment (zimulala) [inline tidb-server/main.go]: If `cluster == ""` it has some problems?

File: tidb-server/main.go [hunk 17]
```diff
@@ -91,14 +98,36 @@ func main() {
...
 	}
+	if *binlogSocket != "" {
+		dialerOpt := grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		})
+		clientCon, err := grpc.Dial(*binlogSocket, dialerOpt, grpc.WithInsecure())
+		if err != nil {
+			log.Fatal(errors.ErrorStack(err))
+		}
+		binloginfo.PumpClient = binlog.NewPumpClient(clientCon)
```
Comment (zimulala) [inline tidb-server/main.go]: Why not check `if binloginfo.PumpClient == nil` here?

File: store/tikv/txn_committer.go [hunk 18]
```diff
@@ -338,35 +341,130 @@ func (c *txnCommitter) Commit() error {
...
 	err = c.commitKeys(NewBackoffer(commitMaxBackoff), c.keys)
 	if err != nil {
 		if !c.mu.committed {
 			log.Warnf("txn commit failed on commit: %v, tid: %d", err, c.startTS)
+			c.rollbackBinlog()
 			return errors.Trace(err)
 		}
 		log.Warnf("txn commit succeed with error: %v, tid: %d", err, c.startTS)
 	}
+	c.writeCommitBinlog()
```
Comment (zimulala) [inline store/tikv/txn_committer.go]: Why not put `c.writeCommitBinlog()` and `c.rollbackBinlog()` in `tikvTxn.Commit()`?

File: store/tikv/txn_committer.go [hunk 19]
```diff
@@ -367,6 +377,89 @@ func (c *txnCommitter) Commit() error {
...
+		prewriteValue := c.txn.us.GetOption(kv.BinlogData)
+		var err error
+		if prewriteValue != nil {
+			binPrewrite := &binlog.Binlog{
+				Tp:            binlog.BinlogType_Prewrite,
+				StartTs:       int64(c.startTS),
+				PrewriteKey:   c.keys[0],
+				PrewriteValue: prewriteValue.([]byte),
+			}
+			prewriteData, _ := binPrewrite.Marshal()
```
Comment (zimulala) [inline store/tikv/txn_committer.go]: Check the error. The same as line 418 and line 442

File: store/tikv/txn_committer.go [hunk 20]
```diff
@@ -367,6 +377,89 @@ func (c *txnCommitter) Commit() error {
...
 
+func (c *txnCommitter) prewriteBinlog() chan error {
+	if !c.shouldWriteBinlog() {
+		return nil
+	}
+	ch := make(chan error, 1)
+	go func() {
+		prewriteValue := c.txn.us.GetOption(kv.BinlogData)
+		var err error
+		if prewriteValue != nil {
```
Comment (zimulala) [inline store/tikv/txn_committer.go]: If the return value of c.shouldWriteBinlog() is ture, the value of `prewriteValue != nil` also is true.

File: store/tikv/txn_committer.go [hunk 21]
```diff
@@ -367,6 +377,89 @@ func (c *txnCommitter) Commit() error {
...
+			binPrewrite := &binlog.Binlog{
+				Tp:            binlog.BinlogType_Prewrite,
+				StartTs:       int64(c.startTS),
+				PrewriteKey:   c.keys[0],
+				PrewriteValue: prewriteValue.([]byte),
+			}
+			prewriteData, _ := binPrewrite.Marshal()
+			req := &binlog.WriteBinlogReq{ClusterID: binloginfo.ClusterID, Payload: prewriteData}
+			var resp *binlog.WriteBinlogResp
```
Comment (zimulala) [inline store/tikv/txn_committer.go]: This line can be removed.

File: sessionctx/binloginfo/binloginfo_test.go [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+		if bin.Tp == binlog.BinlogType_Prewrite {
+			prewriteCount++
+		}
+	}
+	pump.mu.Unlock()
+	match := false
+	for i := 0; i < 10; i++ {
+		pump.mu.Lock()
+		length = len(pump.mu.payloads)
+		pump.mu.Unlock()
```
Comment (zimulala) [inline sessionctx/binloginfo/binloginfo_test.go]: Remove line 183-185.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**